src/main/java/uk/gov/ch/oauth/Oauth2.java

```diff
@@ -241,4 +245,22 @@ private void removeZXSInfo(Map<String, Object> sessionData) {
             sessionFactory.getDefaultStore().delete(zxsKey);
         }
     }
+
+    public String getOriginalRequestURL(final HttpServletRequest request) {
+        final StringBuilder originalRequestUrl = new StringBuilder(
+                request.getRequestURL());
+        final String queryString = request.getQueryString();
+        if (queryString != null) {
+            originalRequestUrl.append("?").append(queryString);
+        }
+        return originalRequestUrl.toString();
+
+    }
+
+    public String prepareState(final HttpServletRequest request) {
+        final Session chSession = (Session) request.getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
```

Comment:
Can we look to get the session in the same way as the other controllers. I would check to see that functionality isn't affected by such a change however.

Reply:
As in get the session in the actual controller, rather than in the oauth2 class?

Reply:
Changed

Reply:
... But I think doing it here rather than the controller makes more sense as it gives us control incase the OAUTH changes rather than making places that implement this have to worry about maintaining it.

Reply:
🤦

Reply:
Yeah sorry the location was fine, I meant as opposed to getting the attribute from request and then casting to a session can we look to get the session like we've done it elsewhere. I.e, using the session factory. However, now I think about it the session factory makes a call to the Handler which essentially does the same thing by getting the servlet request attributes and casting to a session. So we could just leave it like it was, unless there is merit in the semantics or with unit testing (I can't think of any) @CH-JonathanBlake thoughts?

Reply:
My first thought was that we must be doing that because this is a situation where the session in the request is likely to have been altered to the store might not reflect that, this is what we do in the signout journey...
But that doesn't sound like anything we'd expect from the sign in journey.
Where would the request have added a session? Is this something that we do either in our controllers or is it actually likely to be null at this point?
This does seem like the sort of thing that we have the getSessionFromContext() method in the session factory for.

Reply:
So is it fine as it is or do we need to alter parts of the SignInController?

Reply:
Ok, I can try to do that if you think that's the way to go. Currently we don't import SessionFactory into either of the controllers, is a reason why that might not be a good idea.....

Reply:
But we can explain why we do it in the other controller but not here.

Reply:
we don't do it in the other controller....

Reply:
Also when I try to create a sessionFactory field in the signIn controller it wants to be initialized in a constructor. Maybe we should move the session stuff to oauth2 class?

Reply:
Sorry perhaps I was unclear.
In the signout controller we do it this way because we are retreiving session data that has been updated but we don't know has been persisted to the store.
In the signin controller we have no reason to expect this would be case and nobody has yet said where it would even get added to the request.

Reply:
Ok, but then the point remains that I would need to add a constructor to the sign in controller...

Reply:
Why?
`
    public String prepareState(final HttpServletRequest request) {
        final Session chSession = sessionFactory.getSessionFromContext();
        String originalURL = getOriginalRequestURL(request);
        return encodeSignInState(originalURL, chSession, "content");
    }
`
What am I missing that is wrong with that?

Reply:
Right yeah ok, I was confused about the location still

Reply:
Changed